src/math/fft/fft.c

```diff
@@ -0,0 +1,228 @@
+#include <sof/audio/format.h>
+#include <sof/common.h>
+#include <sof/lib/alloc.h>
+#include <sof/math/fft.h>
+
+/*
+ * These helpers are optimized for FFT calculation only.
+ * e.g. _add/sub() assume the output won't be saturate so no check needed,
+ * and _mul() assumes Q1.31 * Q1.31 so the output will be shifted to be Q1.31.
+ */
+
+static void icomplex32_add(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	out->real = in1->real + in2->real;
+	out->imag = in1->imag + in2->imag;
+}
+
+static void icomplex32_sub(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	out->real = in1->real - in2->real;
+	out->imag = in1->imag - in2->imag;
+}
+
+static void icomplex32_mul(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	out->real = ((int64_t)in1->real * in2->real - (int64_t)in1->imag * in2->imag) >> 31;
+	out->imag = ((int64_t)in1->real * in2->imag + (int64_t)in1->imag * in2->real) >> 31;
+}
+
+/* complex conjugate */
+static void icomplex_conj(struct icomplex32 *comp)
+{
+	comp->imag = SATP_INT32((int64_t)-1 * comp->imag);
+}
+
+/* shift a complex n bits, n > 0: left shift, n < 0: right shift */
+static void icomplex_shift(struct icomplex32 *input, int32_t n, struct icomplex32 *output)
+{
+	if (n > 0) {
+		/* need saturation handling */
+		output->real = SATP_INT32(SATM_INT32((int64_t)input->real << n));
+		output->imag = SATP_INT32(SATM_INT32((int64_t)input->imag << n));
+	} else {
+		output->real = input->real >> -n;
+		output->imag = input->imag >> -n;
+	}
+}
+
+/**
+ * \brief Doing Fast Fourier Transform (FFT) or Inverse FFT (IFFT)
+ *	  For input complex buffers.
+ * \param[in] inb - pointer to input buffer.
+ * \param[in] size - input buffer sample count.
+ * \param[in] ifft - set to 1 for IFFT and 0 for FFT.
+ * \param[out] outb - pointer to output buffer.
+ */
+static void fft_internal(struct icomplex32 *inb, struct icomplex32 *outb,
+			 uint32_t size, bool ifft)
+{
+	struct icomplex32 tmp1;
+	struct icomplex32 tmp2;
+	int *bit_reverse_idx;
+	int depth;
+	int lim = 1;
+	int len = 0;
+	int top;
+	int bottom;
+	int index;
+	int i;
+	int j;
+	int k;
+	int m;
+	int n;
+
+	/* calculate the exponent of 2 */
+	while (lim < size) {
+		lim <<= 1;
+		len++;
+	}
+
+	/* convert to complex conjugate for ifft */
+	if (ifft) {
+		for (i = 0; i < size; i++)
+			icomplex_conj(&inb[i]);
+	}
+
+	bit_reverse_idx = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				  lim * sizeof(int32_t));
+
+	/* step 1: re-arrange input in bit reverse order, and shrink the level to avoid overflow */
+	for (i = 1; i < lim; ++i) {
+		bit_reverse_idx[i] = (bit_reverse_idx[i >> 1] >> 1) |
+				     ((i & 1) << (len - 1));
+		icomplex_shift(&inb[i], -len, &outb[bit_reverse_idx[i]]);
+	}
+
+	rfree(bit_reverse_idx);
+
+	/* step 2: loop to do FFT transform in smaller size */
+	for (depth = 1; depth <= len; ++depth) {
+		m = 1 << depth;
+		n = m >> 1;
+		i = FFT_SIZE_MAX / m;
```

Comment:
Avoid division with shift?

Reply:
ack m & n look like powers of 2 so pls shift.